### PR TITLE
Graph Reindexing Issue Fix

### PR DESCRIPTION
**Pull request by shkrishn**
- Fix to avoid index corruption after reindexing RelationTypeIndex with direction IN or BOTH
- While reindexing RelationTypeIndex, edges are created according to the position of relation arity

Steps to reproduce to the self-link issue:
1) Create an Edge label (Assume name is 'link')
2) Create  an property key (Assume name is 'assockind
3) Build an 'Edge Index' by label 'link' and property key 'assockind' with Direction 'IN' or 'BOTH'
4) Create two vertices(let us call it A, B) and add a edge between A and B with label 'link' and some random value (Example : 1) for the property key 'assockind'
5) Query for all the IN edges of vertex A. Make sure the query uses the edge index. This will give no result as there is no IN edge for vertex A
	g.V().has('name' , 'A').inE().hasLabel('link').has('assocKind',1)
6) Reindex the edge index.
	mgmt.updateIndex(index, SchemaAction.REINDEX).get()
7) Execute the Step 3 once again after the reindexing. We can see a self link created for vertex A.

Complete demonstration is provided in below forum post.
https://groups.google.com/forum/#!topic/janusgraph-dev/rZPsk2CDwJ0

Signed-off-by: Shivaa Krishnan <shkrishnan@informatica.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [ x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit? - Single

### For code changes:
- [ x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by farodin91** [comment]
@shkrishn You need to sign the cla.

**Comment by shkrishn** [comment]
> 
> 
> @shkrishn You need to sign the cla.

Hi, Im waiting for the approval from my company side. I ll get it asap.

**Comment by shkrishn** [comment]
> 
> 
> @shkrishn You need to sign the cla.

Hi,

Our company's CLA Signatory has raised an registration request with email id vyedlapalli@informatica.com

It is showing 'Awaiting Approval' since last few days. Could you please help me on this.

Screen Shot 2020-06-29 at 11 49 21 AM

**Comment by farodin91** [comment]
@mbrukman Do you have any idea?

**Comment by mbrukman** [comment]
I've already replied to @shkrishn by email, and @pluradj has involved LF folks to look into this.

FWIW, I think this screenshot shows that it's following the wrong path: they should be associating their account with the Informatica (the company), rather than JanusGraph (the project). Then, once they've established that they represent the company, they need to _sign_ the JanusGraph CCLA, not attempt to become JanusGraph _project managers_ — that's where this is currently stuck.

The docs for this are here:

* [add a company](https://docs.linuxfoundation.org/docs/communitybridge/communitybridge-easycla/ccla-managers-and-ccla-signatories/add-a-company-to-a-project) — note that this should be "Informatica", not "JanusGraph"
* [sign the JanusGraph CCLA](https://docs.linuxfoundation.org/docs/communitybridge/communitybridge-easycla/ccla-managers-and-ccla-signatories/sign-a-corporate-cla-on-behalf-of-the-company) on behalf of Informatica

These steps shouldn't require any approval from LF or JanusGraph.

They're currently stuck waiting for membership in the JanusGraph project, but that will not happen, because we don't add everyone who contributes to JanusGraph as admins/managers of the project on LF.

**Comment by WillsonHG** [comment]
@mbrukman Is correct. I've gone in and removed that Janusgraph "company" listing so that it doesn't cause any more confusion. I see that Informatica was successfully created yesterday, so it seems they are now on the right path to signing the JanusGraph CCLA.

**Comment by shkrishn** [comment]
Reopening after getting the CLA approval

**Comment by shkrishn** [comment]
> 
> 
> @shkrishn You need to sign the cla.

@farodin91 I have got the CLA signed and could see all 3 checks passed. Could you please let me know how to proceed further.

**Comment by shkrishn** [comment]
@pluradj @farodin91 Could you please let me know how to move forward to get this reviewed

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (farodin91) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6409]: seems to be unused.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (farodin91) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6405]: could you use junit5 assertions.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (farodin91) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6294]: Why do you need to run remove?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (farodin91) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:129]:
```suggestion
                Map<StaticBuffer,List<Entry>> inAdditionsMap = new HashMap<>();
```

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (farodin91) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:154]: spacing

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (farodin91) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:161]: indent

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (farodin91) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:142]: indent

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (shkrishn) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6294]: Im creating the vertex with same properties in other testcases also to test index with different directions. Hence removing the vertex after completion of every testcase to avoid conflicts.

Comment (farodin91) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6294]: Tests in this class should be independent.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (farodin91) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:141]: We use space instead of tabs.

Comment (shkrishn) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java]: Replaced all tabs with spaces.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (farodin91) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6292]: You can remove also close.

**Comment by farodin91** [comment]
Could also check the format of the JanusGraphTest seems to be also a bit of.

**Comment by shkrishn** [comment]
> 
> 
> Could also check the format of the JanusGraphTest seems to be also a bit of.



> 
> 
> Could also check the format of the JanusGraphTest seems to be also a bit of.

gG

**Comment by shkrishn** [comment]
Sorry. I accidentally closed the pull request. Reopened it again.

**Comment by shkrishn** [comment]
> 
> 
> Could also check the format of the JanusGraphTest seems to be also a bit of.

Is it about the indent issue. Should I replace all the tabs with spaces?

**Comment by shkrishn** [comment]
Formatting issues fixed.

**Comment by shkrishn** [comment]
Hi @farodin91 , do we have any updates on this request. could you please let me know if anything pending from my side?

**Review by farodin91 — APPROVED** [review]
LGTM! @shkrishn If you like could squash and rebase all commits, otherwise I'll do this on merge.

**Comment by shkrishn** [comment]
Thanks @farodin91. Could you please help me on squashing and rebase commits. I'm not much familiar with this action in GIT.

**Comment by farodin91** [comment]
You could checkout following doc: https://gist.github.com/patik/b8a9dc5cd356f9f6f980

**Comment by mbrukman** [comment]
@farodin91 wrote:

> You could checkout following doc: https://gist.github.com/patik/b8a9dc5cd356f9f6f980

Note that at the top of the doc, it specifically says it's not applicable to PRs with feature branches, and only applicable to folks who have access to push directly to the `master` branch, which is not the case here.

@shkrishn, you probably want to look at https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (li-boxuan) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6374]: `isReindexed` parameter is not used. Anyway, you don't even need a whole method here. I suggest doing
```java
g.V().has("vtName", "A").inE().hasLabel("egLabel").has("assoKind", 1).count().next()
```
directly, wherever you currently use `countOfEdges` method.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (li-boxuan) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6365]: When you put an ancillary method which is for one or more specific scenarios, you'd better not use a seemingly generic name like "assertEdgeCounts".

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (li-boxuan) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6231]: Looks like the three test cases are very similar, except they have different directions. Can you put most of the common logic into one helper method, e.g. "testReindexingForEdgeIndex(direction)", and have 3 very short test cases that use that helper method?

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (li-boxuan) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6350]: Once you extract test logic into a single helper method, e.g. "testReindexingForEdgeIndex(direction)", then you should be able to put the preparation work in that method and dismiss this method.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:151]: IllegalStateException sounds more suitable to me.

**Review by li-boxuan — CHANGES_REQUESTED** [review]
Thanks for the PR! I have some comments, mostly style issues:

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:166]:
```suggestion
                    totalInAdditions += inAdditions.size();
```

**Review by li-boxuan — CHANGES_REQUESTED** [review]

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:163]: Can be just `Map.Entry`, since you have `Map` imported.

**Comment by li-boxuan** [comment]
ping @shkrishn

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (shkrishn) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:147]: code modified.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (shkrishn) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java]: testcase removed as this seems to be unnecessary for this PR

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (shkrishn) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java]: testcase removed as this seems to be unnecessary for this PR

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (shkrishn) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java]: testcase removed as this seems to be unnecessary for this PR

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (shkrishn) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:163]: code modified

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (shkrishn) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java]: code modified

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (shkrishn) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java]: testcase removed as this seems to be unnecessary for this PR

**Comment by shkrishn** [comment]
@farodin91  @li-boxuan Could you please help me why CI Index Solr / tests (solr, -Psolr7) are failing above. This is not related to my changes.

**Comment by li-boxuan** [comment]
I restarted the test and it passed.

**Comment by shkrishn** [comment]
@li-boxuan Thanks for your comments. I have added testcases to repro the issue. Could you please review it.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (li-boxuan) [inline janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6416]: You can do this:
`EdgeLabel edgeLabel = mgmt.makeEdgeLabel("egLabel").multiplicity(Multiplicity.MULTI).make();`
so that you don't need to call `getEdgeLabel` later. Similar for property keys.

**Review by li-boxuan — COMMENTED** [review]
Thanks for adding the test back. I mostly left some stylistic comments.
Can you also do a rebase and squash your commits?

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (shkrishn) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6403]: removed

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (shkrishn) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6415]: done.

File: janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java
Comment (shkrishn) [reply janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java:6416]: done.

**Comment by shkrishn** [comment]
@li-boxuan fixed all stylistic comments. could you please help me in rebase and squash mycommits

`
[root@edcdeployment021 janusgraph]# git log --oneline --author=shkrishn
d3fdbac fixed few stylistic issues
ee4bdac Merge branch 'master' into reindexing-issue-fix
91c7be1 Added refactored testcases
0b732b8 minor code fix
4f8d625 Merging with master
457f0a3 Formatting fix
6306a62 removed close() call
b17c503 Replaced tabs with spaces
5ea2b19 Merge branch 'master' into reindexing-issue-fix
27abbde Fixing indent issues
f618bc4 Graph Reindexing Issue Fix
`

**Comment by shkrishn** [comment]
@li-boxuan  fixed all stylistic comments. could you please help me in rebase and squash mycommits

**[root@edcdeployment021 janusgraph]# git log --oneline --author=shkrishn**
d3fdbac fixed few stylistic issues
ee4bdac Merge branch 'master' into reindexing-issue-fix
91c7be1 Added refactored testcases
0b732b8 minor code fix
4f8d625 Merging with master
457f0a3 Formatting fix
6306a62 removed close() call
b17c503 Replaced tabs with spaces
5ea2b19 Merge branch 'master' into reindexing-issue-fix
27abbde Fixing indent issues
f618bc4 Graph Reindexing Issue Fix

**Comment by farodin91** [comment]
@shkrishn https://dev.to/victoria/a-quick-guide-to-squashing-git-commits-7io

I think that article helps a lot.

**Comment by li-boxuan** [comment]
@shkrishn We can help you do the rebase and squash, but trying it yourself would be a good learning experience (I am assuming you haven't done git rebase & squash before). Let us know if you need any help.

**Comment by li-boxuan** [comment]
@shkrishn If you have difficulty doing squash/rebase, the easiest way is to copy your latest changes, reset your branch to the last commit on master before your first commit, apply your latest changes and do a new single commit.

**Comment by shkrishn** [comment]
@li-boxuan  Sorry for the late response. Got stuck with some personal work. I have reset my branch to the last commit on master, applied my latest changes and did a new single commit.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java
Comment (li-boxuan) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:149]: I find that your test case testReindexingForEdgeIndex will pass even if I remove this whole "else" branch. There are two possibilities:
1) Your test case is not effective enough to cover the edge case
2) This is actually not needed. Maybe only the case where pos == 0 is needed.
Can you please investigate?

Comment (shkrishn) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:149]: @li-boxuan  Sure i ll look into this issue.

Comment (li-boxuan) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java:149]: Your fix is correct. It's just that your test case is not effective enough. I can provide a subsequent PR to improve the tests.

**Review by li-boxuan — APPROVED** [review]
The fix looks good, thank you!

**Comment by farodin91** [comment]
@li-boxuan Sounds great. I will merge it.

**Comment by shkrishn** [comment]
Thanks all for merging this :)

**Pull request merged**
